### PR TITLE
Implement Delete Appointment command

### LINKED ISSUE #71: Implement "Delete Appointment" Feature
**Feature**: Delete Appointment

**Purpose**: Allows doctors to delete an appointment with a patient. 

**Command format**: `delAppt [i/NRIC] [d/DATE] [t/TIME]`

**Example commands**:
`delAppt i/T0123569I d/2024-09-20 t/1000-1100`

**Acceptable values**:
- NRIC: 
  - Must be 9 characters
  - first and last characters being alphabets
  - Follows the 9-character Singapore NRIC format for uniqueness.
  - Case-insensitive (e.g. ‘TXXXXXXXB’ is the same as ‘tXXXXXXXb’) for consistency.
- Date: 
  - Must be in the YYYY-MM-DD format (e.g., "2024-09-20").
  - Standard YYYY-MM-DD format for consistency and scheduling ease.
- Time: 
  - Must be in the HHMM format (24-hour clock) (e.g., "1000") 
  - must be in an interval (e.g. 1000-1100). 
  - Appointment time intervals cannot overlap with other existing appointment time intervals.
  - Standard HHMM format for consistency and scheduling ease.
  - Time format should be (start time - end time) where start time < end time.
  - The start time should not equal the end time. 

**Error message if invalid**:
- "Invalid date format. Use YYYY-MM-DD."
- "Invalid time format. Use HH:MM."
- "Patient does not exist.”
- “There is no valid appointment for this date and time”

**Outputs**:
- Success: "Appointment for John Doe canceled on 2024-09-20 at 1700-1900."
- Failure: Error message related to the incorrect input format.
- Null handling: The system will check if an appointment does not exist for the given date and time, prompting a message like "Appointment does not exist for this date and time."

### PR DESCRIPTION
**Pull request by yadobler**
Fixes #71 

oppenheimer-cillian-murphy

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/DeleteApptCommandParser.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    public DeleteApptCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NRIC, PREFIX_DATE,
+                                                                  PREFIX_TIMEPERIOD);
+        String patientNricValue;
+        String newApptDate;
+        String newApptTime;
+
+        Supplier<ParseException> parseException = () -> new ParseException(
+                String.format(MESSAGE_INVALID_COMMAND_FORMAT, MESSAGE_USAGE));
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/DeleteApptCommandParser.java:40]: Good use of the `Supplier` class

File: src/main/java/seedu/address/model/person/AppointmentExistsPredicate.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public boolean test(Person person) {
+        apptMatch = person.getAppointments()
+                          .stream()
+                          .filter(x -> x.getAppointmentDate().equals(date))
+                          .filter(x -> x.getAppointmentTimePeriod().equals(timePeriod))
+                          .findFirst();
+        return apptMatch.isPresent();
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/model/person/AppointmentExistsPredicate.java:38]: SLAP is adhered to. Nice ⚡

File: src/test/java/seedu/address/logic/commands/AddApptCommandTest.java [hunk 3]
```diff
@@ -28,34 +30,47 @@
 import seedu.address.testutil.PersonBuilder;
 
 /**
- * Contains the integration tests (interaction with the Model) and unit tests for AddApptCommand.
+ * Contains the integname.equals(otherPerson.name)ration tests (interaction with the Model) and unit tests for
+ * AddApptCommand.
```
Comment (yong-zhuo) [inline src/test/java/seedu/address/logic/commands/AddApptCommandTest.java:34]: Remember to fix this javadocs header comment

File: src/test/java/seedu/address/logic/commands/DeleteApptCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    public void execute_allFieldsSpecified_success() {
+        Model model = new ModelManager(getTypicalAddressBook(), new UserPrefs());
+        Person targetPerson = model.getFilteredPersonList().get(0);
+        Nric nric = targetPerson.getNric();
+        Appointment apptToDelete = targetPerson.getAppointments()
+                                               .stream()
+                                               .findFirst()
+                                               .orElseThrow(() -> new AssertionError(
+                                                       DeleteApptCommandParserTest.class
+                                                       + ": PLEASE USE ANOTHER TARGET")
```
Comment (yong-zhuo) [inline src/test/java/seedu/address/logic/commands/DeleteApptCommandTest.java:45]: Just a nit: Could be possible to use `Supplier` to abstract out  
`() -> new AssertionError(DeleteApptCommandParserTest.class
                                                       + ": PLEASE USE ANOTHER TARGET"`
                                                       
                                                       
But test cases for this class is comprehensive 👍🏻

**Review by yong-zhuo — APPROVED** [review]
LGTM overall, no major issues with the code.

File: src/main/java/seedu/address/logic/commands/DeleteApptCommand.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    private final NricMatchesPredicate targetNric;
+    private final AppointmentExistsPredicate targetAppt;
+
+    /**
+     * Constructs the command for delAppt.
+     *
+     * @param targetNric Predicate for testing existing {@code Nric}.
+     * @param targetAppt Predicate for testing existing {@code Appointment}.
+     */
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/commands/DeleteApptCommand.java:66]: Neat javadocs! good job!

File: src/main/java/seedu/address/logic/commands/DeleteApptCommand.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+        Person personToEdit = model.fetchPersonIfPresent(targetNric)
+                                   .orElseThrow(() -> new CommandException(Messages.MESSAGE_PERSON_NRIC_NOT_FOUND));
+        Appointment apptToDelete = targetAppt.getApptMatch(personToEdit)
+                                             .orElseThrow(() -> new CommandException(MESSAGE_PERSON_APPT_NOT_FOUND));
+        Person editedPerson = createEditedPerson(personToEdit, apptToDelete);
+
+        model.setPerson(personToEdit, editedPerson);
+        logger.info("delAppt command has been exceuted successfully on a person with NRIC: " + targetNric + " for "
+                    + "appointment: " + targetAppt);
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/commands/DeleteApptCommand.java:86]: Good use of logger.

File: src/main/java/seedu/address/logic/parser/DeleteApptCommandParser.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import seedu.address.commons.util.DateUtil;
+import seedu.address.logic.commands.DeleteApptCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.model.appointment.Appointment;
+import seedu.address.model.person.AppointmentExistsPredicate;
+import seedu.address.model.person.Nric;
+import seedu.address.model.person.NricMatchesPredicate;
+
+/**
+ * Parses input arguments and creates a new DeleteApptCommand object
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/parser/DeleteApptCommandParser.java:21]: Javadocs require punctuation

File: src/test/java/seedu/address/logic/commands/DeleteApptCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        Person targetPerson = model.getFilteredPersonList().get(0);
+        Nric nric = targetPerson.getNric();
+        Appointment apptToDelete = targetPerson.getAppointments()
+                                               .stream()
+                                               .findFirst()
+                                               .orElseThrow(TESTCASE_WITHOUT_APPOINTMENT
+        );
+        DeleteApptCommand delApptCommandWrongDate = new DeleteApptCommand(
+                new NricMatchesPredicate(nric),
+                new AppointmentExistsPredicate("1900-01-01", apptToDelete.getAppointmentTimePeriod()));
```
Comment (hweinian) [inline src/test/java/seedu/address/logic/commands/DeleteApptCommandTest.java:110]: Just a nit: string literal could be assigned to a variable first

File: src/test/java/seedu/address/logic/commands/DeleteApptCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                                                                              apptToDelete.getAppointmentTimePeriod());
+        DeleteApptCommand delApptPersonCommand = new DeleteApptCommand(nricNotFound, validAppt);
+        assertCommandFailure(delApptPersonCommand, new ModelManager(getTypicalAddressBook(), new UserPrefs()),
+                             Messages.MESSAGE_PERSON_NRIC_NOT_FOUND);
+    }
+
+    @Test
+    public void execute_patientMissingAppointment_failure() {
+        Model model = new ModelManager(getTypicalAddressBook(), new UserPrefs());
+        Person targetPerson = model.getFilteredPersonList().get(0);
```
Comment (hweinian) [inline src/test/java/seedu/address/logic/commands/DeleteApptCommandTest.java:101]: Just a nit: I think the 0 is a code dupe. There is a constant called ZERO_INDEX

**Review by hweinian — COMMENTED** [review]
Overall LGTM! Most of the stuff is a nit

File: src/main/java/seedu/address/model/person/AppointmentExistsPredicate.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public AppointmentExistsPredicate(String date, String timePeriod) {
+        this.date = date;
+        this.timePeriod = timePeriod;
+        this.apptMatch = Optional.empty();
+    }
+
+    @Override
+    public boolean test(Person person) {
+        apptMatch = person.getAppointments()
```
Comment (TechRaj) [inline src/main/java/seedu/address/model/person/AppointmentExistsPredicate.java:33]: Omg this looks so clean. Nice one!

File: src/test/java/seedu/address/logic/commands/AddApptCommandTest.java [hunk 11]
```diff
@@ -36,26 +38,38 @@ public class AddApptCommandTest {
     @Test
     public void execute_allFieldsSpecified_success() {
         Model model = new ModelManager(getTypicalAddressBook(), new UserPrefs());
-        Nric nric = (model).getFilteredPersonList().get(0).getNric();
-        AddApptCommand addApptCommand = new AddApptCommand(new NricMatchesPredicate(nric), VALID_APPOINTMENT_NAME_AMY,
+        Model expectedModel = new ModelManager(getTypicalAddressBook(), new UserPrefs());
+
+        Person targetPerson = model.getFilteredPersonList().get(0);
```
Comment (TechRaj) [inline src/test/java/seedu/address/logic/commands/AddApptCommandTest.java:43]: Good use of person class to make the code easier to read and easier to debug

File: src/test/java/seedu/address/logic/commands/DeleteApptCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+    public static final Supplier<AssertionError> TESTCASE_WITHOUT_APPOINTMENT = () -> new AssertionError(
+            DeleteApptCommandParserTest.class + ": PLEASE USE ANOTHER TARGET");
+
+    @Test
+    public void execute_allFieldsSpecified_success() {
+        Model model = new ModelManager(getTypicalAddressBook(), new UserPrefs());
+        Person targetPerson = model.getFilteredPersonList().get(0);
+        Nric nric = targetPerson.getNric();
+        Appointment apptToDelete = targetPerson.getAppointments()
```
Comment (TechRaj) [inline src/test/java/seedu/address/logic/commands/DeleteApptCommandTest.java:45]: Good use of streams!

**Review by TechRaj — COMMENTED** [review]
LGTM. Good job! Looks neat!
super-rajini

**Pull request merged**